---
title: "Example code for single-cell analysis with Seurat, day 1"
author: "Kevin Rue-Albrecht"
date: "05/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)
```

# Exercise

## Import scRNA-seq data and create a Seurat object

- Load the `Seurat` package.

```{r}

```

- Use the function `Read10X()` to import data in the directory `filtered_feature_bc_matrix/` 
  as an object named `read10x_data`.
  What class of object does the function return?

```{r}
read10x_data <- Read10X("/project/obds/shared/resources/4_r_single_cell/singlecell_seuratday1/filtered_feature_bc_matrix")
#NB, use gene.column = 1 to use the Ensembl ID instead of gene symbols.
```

```{r}
class(read10x_data)
#Matrix - it is not a suerat object initially
```

> Answer:
> 

- Have a look at the object and its structure (e.g., first 15 rows and 6 columns).
  What is a sparse matrix and how does it store data?

```{r}
dim(read10x_data)

read10x_data[1:15, 1:6] #all the dots (.) represent zeros. This is to save memory. it is a spare data set. you need to store 3 values then thoguh, which row, which column, and which value. By including only this info for non-zero values, you save space. 
# The read function, if there are genes with the same name, will add .1 or .2 etc.if it appears multiple times. Why does this happen? it may happen if the same gene appears multiple times on the same chromosome or appears on multiple chromosomes. 
```

```{r}

```

> Answer:
> 

- How many features and barcodes (i.e., cells) are present in the data set?

```{r}
#dim achieves this
dim(read10x_data)
#33000genes (genes are features), 5000 cells
```

> Answer:
> 

- Create a Seurat object using the function `CreateSeuratObject()`
  and the object `read10x_data`.
  Name the object `seurat_object`.
  Include features detected in at least 3 cells,
  and cells where at least 200 features detected.
  Name the project `pbmc5k`.
  How many features and barcodes are left in the Seurat object?

```{r}
#seurat considers one project one dataset. 
seurat_object <- CreateSeuratObject(
    read10x_data, 
    project = "pbmc5k", #this is arbitrary and is used to assign a name to the cells
    min.features = 200, #this is an arbitrary value we've selected to show that we can specify
    min.cells = 3 #this is an arbitrary value we've selected to show that we can specify
    )
seurat_object


```

> Answer:
> 

- How many features and cells were filtered out when you created the Seurat object?

```{r}
dim(read10x_data) - dim(seurat_object)
#seurat objects act like a matrix. so you can compare the objects before and after filtering to see how many things you have discarded. 
# we say barcode instead of cells right now, because there may be doublets. cell ranger removes empty droplets. so when do you start calling things cells? towards the end of your analysis. After you've done your filtering. each column in these matrices are barcodes. when we start putting them in UMAPs etc. they should be cells at that point so we refer to them as cells
#the project name is a label you put on the whole object saying e.g. this is my sample 1 data set. e.g. if you sample 4 patients, each patient can be a dataset. 
```

> 

# Exercise

## Accessing the contents of a Seurat object

- Query the name of the default assay in the Seurat object.

```{r}
class(seurat_object)

DefaultAssay(seurat_object)
# answer = RNA

#you can aslo do: seurat_object@active.assay - this directly accesses the active assay from within the object. but if a function exists it is always best to use the function. 
```

- List the names of assays available in the Seurat object.

```{r}
# only one assay is available, called RNA
Assays(seurat_object) #returns you just the name

# seurat_object@assays$RNA@data is an alternative. 

```

- Display the first six rows and six columns of the RNA assay data.
  What function do you use?
  Which arguments do you need to supply, and which ones are optional?

```{r}
data <- GetAssayData(seurat_object, slot = "data")
data[1:6, 1:6]
#there's only one assay, but before we normalise, the raw data is copied to the normalised folder. this is why we're still able to access things using the data slot. 

#Can also be done in a single line 
data <- GetAssayData(seurat_object, slot = "data")[1:6, 1:6]

#seurat_object@assays$RNA@data[1:6, 1:6]
#can also do GetAssaysData(seurat_object, assay = "RNA", slot = "data")

```

> Answer:
> 

- Display the entire data.frame of per-cell metadata (first six rows).
  What column names do you see?

```{r}
seurat_object[[]][1:6, ]

# you can also do: head(seurat_object[[]])
```

> Answer:
> 

- Fetch one column of metadata using `[[`.
  What type of object do you get back?

```{r}
seurat_object[["nCount_RNA"]]

```

> Answer:
> 

- Instead,fetch the same column of metadata using `$`.
  What type of object do you get back this time?

```{r}
#try using dollar sign instead of square brackets
seurat_object$nCount_RNA

#best to run head around things when you know it will be long
head(seurat_object$nCount_RNA)
```

> Answer:
> 

- Use the function `FetchData()` to access the library size and expression of the feature named `"LYZ"` (first six rows).
  What type of object do you get back?

```{r}
FetchData(seurat_object, vars = c("LYZ", "nCount_RNA"), slot = "data")[1:6, ]
#key poitn is that we can take things from different areas of the obejct - one from the counts bit and one from the row names of the assay slot! 
```

> 

# Demo

## Common operations on Seurat objects

`WhichCells()` returns the names of cells that match a logical expression.

```{r}
WhichCells(seurat_object, expression = LYZ > 500)
```

`VariableFeatures()` returns the names of variable features (for a given assay, if computed).

```{r}
VariableFeatures(seurat_object)
```

`subset()` returns a new Seurat object restricted to certain features and cells.

```{r}
subset(
    x = seurat_object,
    cells = WhichCells(seurat_object, expression = LYZ > 500),
    features = VariableFeatures(object = seurat_object)
)
```

# Exercise

## Quality control and visualisation

- The library size and number of features detected per cell is already present in the Seurat object.
  Use the function `VlnPlot()` to display them in a single violin plot.

```{r}
VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
```

- Use the function `PercentageFeatureSet()` to compute the fraction of reads
  assigned to mitochondrial genes in each cell.
  Store the metric in the cell metadata of the Seurat object, under the name `"percent_mt"`.
  Visualise this new metric alongside the previous two in a new violin plot.

```{r}
seurat_object[["percent_mt"]] <- PercentageFeatureSet(object = seurat_object, pattern = "MT-")
#need to assign the pattern to a new column in the object.

VlnPlot(seurat_object, features = c("nFeature_RNA", "nCount_RNA", "percent_mt" ), ncol = 3)

```

- Visualise a scatter plot of the proportion of mitochondrial UMIs against the library size in each cell.

```{r}
#ggplot(data = seurat_object) + 
#  geom_dotplot(aes(x = "nCount_RNA", y = "percent_mt"))
#this was wrong bc it wasn't a dataframe, could extract relevant cols into a dataframe then use ggplot

FeatureScatter(seurat_object, 'nCount_RNA', 'percent_mt')




```

- Create a new Seurat object, called `seurat_after_qc`, that is subsetted to cells that have more than 4,500 UMI counts, less than 15% of UMI counts assigned to mitochondrial features, and more than 1,500 features detected.
  How many cells were removed in this step?

```{r}
seurat_after_qc <- subset(
    seurat_object, 
    subset = nCount_RNA > 4500 & percent_mt < 15 & nFeature_RNA > 1500
)
seurat_after_qc
```

```{r}
dim(seurat_object) - dim(seurat_after_qc)
```

> Answer:
> 

# Exercise

## Normalisation

- Normalise the RNA assay of the Seurat object (after quality control) using the `"LogNormalize"` method.

```{r}
seurat_after_qc <- NormalizeData(seurat_after_qc, normalization.method = "LogNormalize")
```

## Bonus

- Visualise the distribution of raw counts and normalised data for a feature of your choice.

```{r}


#FetchData(seurat_object, vars = c("LYZ", "nCount_RNA"), slot = "data")
#FetchData(seurat_object, vars = c("LYZ", "nCount_RNA"), slot = "counts")


#GetAssayData(seurat_after_qc, slot = "counts")
#GetAssayData(seurat_after_qc, slot = "data")


ggplot_lyz_raw <- ggplot(FetchData(seurat_after_qc, vars = c("LYZ", "nCount_RNA"), slot = "counts"), aes(x = LYZ)) +
    geom_histogram() +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
ggplot_lyz_normalised <- ggplot(FetchData(seurat_after_qc, vars = c("LYZ", "nCount_RNA"), slot = "data"), aes(x = LYZ)) +
    geom_histogram() +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised, ncol = 1)



```

# Exercise

## Variable features and scaling

- Identify variable features in the normalised RNA assay of the Seurat object.
  Use the `"vst"` method and select the 2,000 most variable features.
  What does this subsetting do, and what are our motivations for doing it?

```{r}
seurat_after_qc <- FindVariableFeatures(seurat_after_qc, selection.method = 'vst', nfeatures = 2000)
```

> Answer:
> 

- What is the function to display the name of variable features in a Seurat object (e.g., first 10)?
  How can you control which assay the variable features are pull from?

```{r}
VariableFeatures(seurat_after_qc)[1:10]
```

> Answer:
> 

- Use the function `VariableFeaturePlot()` to visualise the scatter plot of standardised variance against average expression.
  How would you use this plot?

```{r}
VariableFeaturePlot(seurat_after_qc)
```

> Answer:
> 

- Scale the normalised RNA assay of the Seurat object, regressing the library size and the fraction of UMI counts assigned to mitochondrial features.
  What are the motivations for removing those two sources of variation?

```{r}


seurat_after_qc <- ScaleData(
  seurat_after_qc, 
  vars.to.regress = c("nCount_RNA", "percent_mt")
)

#head(GetAssayData(seurat_after_qc, slot = 'scale.data'))
GetAssayData(seurat_after_qc, slot = 'scale.data')[1:6, 1:6] #this is helpful - we can check that the values look like they belong on a standard normal distribution

```

> Answer:
> 

# Exercise

## Dimensionality reduction

- Run a principal component analysis on the Seurat object.
  Which features are used by the method in the default settings?
  How could you change this?
  How do you read the message output of the function `RunPCA()`?

```{r}
seurat_after_qc <- RunPCA(seurat_after_qc)
```

> Answer:
> 

- List the names of dimensionality reduction results available in the Seurat object.

```{r}
Reductions(seurat_after_qc)
```

- Use `PCAPlot()` or `DimPlot()` to produce a scatterplot of the first and second PCA components.

```{r}
PCAPlot(seurat_after_qc)

DimPlot(seurat_after_qc, reduction = 'pca')
```

## Bonus

- Make a scatterplot of the first and second PCA components yourself using `ggplot2`.

```{r}
# Use this code chunk to prepare a data.frame for ggplot2
pca_data <- FetchData(seurat_after_qc, vars = c('PC_1', 'PC_2'))
head(pca_data)
```

```{r}
ggplot(pca_data, aes(x = PC_1, y = PC_2))+
    geom_point(size = 0.2) +
    cowplot::theme_cowplot()
```

- Visualise the amount of variance explained the top principal components (number of your choice).
  How many principal components would you use for downstream analyses?

```{r}
ElbowPlot(seurat_after_qc, ndims = 50, reduction = 'pca') +
  coord_cartesian(ylim = c(0, 15))
#ndims is how many dimensions it will display. remember there will be as many principle components as samples. you may need to increase it from the default of 20, which may help you 
```

> 

- Run the UMAP technique on your selected number of principal components and visualise the result as a scatterplot.

```{r}
set.seed(1)
seurat_after_qc <- RunUMAP(seurat_after_qc, dims = 1:18, reduction = "pca")
UMAPPlot(seurat_after_qc)

saveRDS(seurat_after_qc, file = "seurat_after_qc_20Feb2023.RDS")
```

# Exercise

## Clustering

- Compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of principal components that you have chosen earlier.

```{r}
seurat_after_qc <- FindNeighbors(seurat_after_qc, reduction = "pca", dims = 1:18, k.param = 20)
```

> Answer:
> 
> The help page states that the function `FindNeighbors()` uses principal components 1 through 10, by default.

- What are the names of the nearest neighbour graphs that are now stored in the Seurat object?

```{r}
Graphs(seurat_after_qc)
```

- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?

```{r}
seurat_after_qc <- FindClusters(seurat_after_qc, resolution = 0.3)

#more or fewer clusters: 

seurat_after_qc[[]]
```

> 

- Visualise the cluster labels on the UMAP scatter plot.
  How would you describe the agreement between the UMAP layout and the clustering results?

```{r}
DimPlot(seurat_after_qc, label = TRUE, group.by = "RNA_snn_res.0.3")

#can use clustree at this point, a good package. can also use many different resolution values, and then try to find the sweet spot. 
#it might be good to make a panel of figures here with values of resolution from e.g. 0.3 to 1, and compare them all. you can run one resolution at a time. it is more efficient to run it as a vector in a single line. 
```

> 

# Exercise

## Identify cluster markers

- Use the function `FindAllMarkers()` to identify
  positive markers for all clusters,
  filtering markers that are detected in at least 25% of the cluster,
  and with a log fold-change greater than `0.25`.
  Assign the result to an object named `seurat_markers_all`.
  What is the class of that object?
  How do you control the set of clusters that are used?

```{r}
#if you, at this point, don't find good markers, you're data may still be fine, you might just need to go back and change things, e.g. choose different number of PCs, etc. 

#Notes. the DimPlot gives the first graph on slide 48, feature plot gives the second graph. do heat map gives the heatmap on slide 49

#this line of code allows you to specify the column to be used in the next function, find all markers
Idents(seurat_after_qc) <- "RNA_snn_res.0.5"

seurat_markers_all <- FindAllMarkers(seurat_after_qc, 
  logfc.threshold = 0.25, 
  min.pct = 0.25
)

head(seurat_markers_all)
class(seurat_markers_all)


#use a for loop or an lapply to run the function multiple times (lapply you would need to make sure you write to a separate output file each time). this would allow you to find markers using different thresholds for fold change level and proportion of cells expressing it. you need to be careful, you might be too harsh with your criteria and then not get enough data
```

> Answer:
> 

- How do you read the contents of the object `seurat_markers_all`?
  How do you know which features are the markers of each cluster?

```{r}
head(seurat_markers_all)
```

> Answer:
> 

- Filter and display the top 10 markers for cluster 3.

```{r}
#top by fold change or p-value? it's up to you! 

seurat_markers_all %>% filter(cluster == 3) %>% 
  arrange(desc(abs(avg_log2FC))) %>% 
  slice_head(n=10) -> top_10_markers_for_3 #can assign to a variable at the end of the pipe!


top_10_markers_for_3

top_10_markers_for_3_vector = pull(top_10_markers_for_3, gene) #make it into a vector so that you can feed it into the feature plot function. 
```

- Visualise the expression of the top 4 marker for cluster 3 on a UMAP layout.

```{r}
FeaturePlot(seurat_after_qc, 
            features = top_10_markers_for_3_vector[1:4],
            label = TRUE)

#NB, if you delete abs from the workflow above, you get the postivie expression 
```

- Visualise the expression of those same 4 marker as a violin plot.
  Do you have any particular preference between the two types of plots?

```{r}
VlnPlot(seurat_after_qc, 
            features = top_10_markers_for_3_vector[1:4]
)
```

> Answer:
> 

- Use `DoHeatmap()` to visualise the top 10 (positive) markers for each cluster.
  Hint: you may want to use the function `dplyr::group_by()`.

```{r} 
markers_top10_clusters <- seurat_markers_all %>% group_by(cluster) %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% slice_head(n=10) %>% pull(gene)
    
#NEED TO ADD A FILTER FOR P VALUES AT SOME POINT bc now we're just looking at fold change, there might be non-significant values. 
    
DoHeatmap(seurat_after_qc, features = markers_top10_clusters, size = 3) #might need to change figure height and width
```

