---
title: Week4-day5 :integration template"
author: "Devika Agarwal/updataed by Carla Cohen"
date: "17/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) #print code and results
```


```{r}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(patchwork)
library(tidyverse)
library(batchelor)
library(harmony)
library(lisi)
library(MAST)
options(future.globals.maxSize = 4000 * 1024^2)    # Maximum allowed total size (in bytes) of global variables identified
```

# Exercise 1 - joint analysis of two PBMC datasets

Here we will look at how to PBMC datasets compare without any integration. 

## 1. Import the two RDS files into R

These RDS object have already been put through the QC and filtering steps covered on Seurat day 1.


```{r}

?readRDS

pbmcv2 <- readRDS("/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv2_filtered.rds")
pbmcv3 <- readRDS("/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv3_filtered.rds")
```


Plot these metadata covariates nCount_RNA/nFeature_RNA/percent.mt to see how they range for both the datasets 

use `VlnPlot()` from Seurat

```{r}
pbmcv2
pbmcv2[[]] #to see  metadata col names
?VlnPlot

p1 <- VlnPlot(pbmcv2, features = c('nCount_RNA', 'nFeature_RNA', 'percent.mt'))
p2 <- VlnPlot(pbmcv3, features = c('nCount_RNA', 'nFeature_RNA', 'percent.mt'))
cowplot::plot_grid(p1, p2, ncol = 1)
```



## 2. Create a merged Seurat object containing both sets of data (no batch correction)

Vignette: https://satijalab.org/seurat/articles/merge_vignette.html

```{r}
seurat_merged <- merge(x = pbmcv2, y = pbmcv3, add.cell.ids = c('v2', 'v3'))
seurat_merged[[]] #you can see that it has renamed each of the cell barcodes with the 
dim(seurat_merged)
#the data label lives within the orig idents column. 
```

Check the dims of the merged object in comparison to the separate objects

What is the total gene number in the merged object compared to each single object?

```{r}
dim(seurat_merged)
dim(pbmcv2)
dim(pbmcv3)
```

## 3.1 Perform SCTransform normalisation

Do you want to regress any meta.data variables ?

```{r, warning=FALSE}
seurat_merged <- SCTransform(seurat_merged, assay="RNA", new.assay.name = "SCT", vars.to.regress ="percent.mt")

```

Plot to see how the nCount_RNA looks before and after SCT normalisation for the merged dataset?

Did SCT do a good job in normalizing the sequencing depth across the two datasets?


```{r}

p1 <- VlnPlot(seurat_merged, features = c('nCount_RNA', 'nCount_SCT'), same.y.lims =TRUE, y.max = 12000)
p1
seurat_merged[[]]

#scale_y_continuous() ???
```


What should the Default Assay be?

```{r}
DefaultAssay(seurat_merged)

#sct, because the sct step converts the default assay to SCT assay
```

Plot the Variable features plot for data with top 10 variable genes labelled

```{r, fig.width=4}

top_10_var_features = VariableFeatures(seurat_merged)[1:10]

variable_feature_plot <- VariableFeaturePlot(seurat_merged, selection.method = "SCT")

?LabelPoints

LabelPoints(variable_feature_plot, points = top_10_var_features, repel = TRUE, ynudge = 0, xnudge = 0)

```

## 3.2 Run PCA and visualize 

Can you see a batch effect between pbmc v2 and v3?


```{r}

seurat_merged <- RunPCA(seurat_merged, assay = "SCT", npcs = 50)
```

Look at which genes are driving your PCs.

```{r, fig.width=4}
VizDimLoadings(seurat_merged, dims = 1:2, nfeatures = 20, balanced = TRUE)
```


```{r}
?DimPlot
DimPlot(seurat_merged, reduction = "pca")


```

Choose optimal  number of PCs for UMAP

```{r}

?ElbowPlot
ElbowPlot(seurat_merged, ndims = 50, reduction = "pca")
#cut off at 20 dimensions
```

## 3.3 Run the Seurat UMAP function and visualize

Is there a batch effect? 


```{r}
?RunUMAP
seurat_merged <- RunUMAP(object = seurat_merged, dims = 1:20, reduction = "pca", assay = "SCT")

Reductions(seurat_merged) #to find name of the umap reduction to enter into dimplot

DimPlot(seurat_merged, reduction = "umap")
```

### Optional: Do cells cluster by celltypes?

Plot some pbmc celltype markers to see how they look on the UMAP

```{r}
?FeaturePlot
FeaturePlot(seurat_merged, features = c("CD3D", "CD3E", "GNLY", "LYZ", "S100A8", "S100A9", "CD79A", "CD79B", "MS4A1"), reduction = "umap")
```


## Exercise 2: Seurat integration with SCTransform

Check out the Vignette: https://satijalab.org/seurat/articles/integration_introduction.html#performing-integration-on-datasets-normalized-with-sctransform-1 

The Seurat v2 integration method is slightly different depending on whether you perform log normalistaion or SCTransformation on your seurat objects. Here we are using the SCTransform method (the vignette covers both).

### 1. Set up

Import the RDS files 

We will read in the objects again to start the integration workflow from scratch

```{r}
pbmcv2 <- readRDS("/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv2_filtered.rds")
pbmcv3 <- readRDS("/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv3_filtered.rds")
```

Generate a named list of the two objects to run workflow

```{r}
pbmc.list <- list("v2" = pbmcv2, "v3" = pbmcv3)
pbmc.list

#this step is in contrast to the merge step we performed previously 
```

Apply SCTransform across the list

```{r}

?lapply
pbmc.list <- lapply(pbmc.list, SCTransform, assay="RNA", new.assay.name = "SCT", vars.to.regress ="percent.mt")
#previously we performed SCT on all the cells at once. now we're doing it on two separate data sets.Here, we're performing the SCT on the two objects separately. As such, we'll get two separate lists of variable gene expression 

#notes for lapply fx - the ... in the documentation shows the space where you put the arguments belonging to the the function that occupies the FUN argument. 
```

Check dims for both objects in list after SCTransform
```{r}

#pbmc.list["v2"]

dim(pbmc.list[["v2"]])
dim(pbmc.list[["v3"]])

#or
dim(pbmc.list$v2)
dim(pbmc.list$v3)

```

Make sure Default assay in both objects is SCT
```{r}
?DefaultAssay
DefaultAssay(pbmc.list$v2)
#if it wasn't sct, reassign by: 
DefaultAssay(pbmc.list$v2) <- "SCT"

#checking the other item in the list:
DefaultAssay(pbmc.list$v3)

#being funky and doing a single lapply 
?lapply
lapply(pbmc.list, DefaultAssay)

```

Choose the features to use when integrating multiple datasets. 

This function ranks features by the number of datasets they are deemed variable in, breaking ties by the median variable feature rank across datasets. It returns the top scoring features by this ranking.
```{r}
?SelectIntegrationFeatures
features <- SelectIntegrationFeatures(object.list = pbmc.list, assay = c("SCT", "SCT"), nfeatures = 2000) 
features #Here, we have made a vector of integration features, i.e. highly variable genes across both datasets. 
```

Prepare an object list normalized with sctransform for integration (run `PrepSCTIntegration()` )

This function subsets the scale.data slot to only contain residuals for anchor.features for efficiency in downstream processing
It is specific to the SCTransform workflow and makes sure you have the SCTransform residuals for the remaining genes

```{r}
?PrepSCTIntegration
pbmc.list <- PrepSCTIntegration(pbmc.list, assay = "SCT", anchor.features = features)

```

> After this step, the features present in the scale.data slots of the two objects are the same
> anchor.features are used in the anchor finding process


### Optional exercise

How similar are these features to the variable features that we found for the pbmcv2 and v3 individually?
- extract variable features for each dataset individually
- compare these with "features"; which has the variable features most common for both datasets


## 2. Perform integration

Find a set of anchors between a list of Seurat objects. These anchors can later be used to integrate the objects.

```{r}
?FindIntegrationAnchors
pbmc.anchors <- FindIntegrationAnchors(pbmc.list, normalization.method = "SCT", anchor.features = features, dims = 1:20)

```

Integrate the two pbmc objects

```{r}
?IntegrateData
pbmc.combined.sct <- IntegrateData(anchorset = pbmc.anchors, new.assay.name = "v2_v3_integrated", normalization.method = "SCT", dims = 1:20) #NB, the SCT here is not performing any SCTs, it's simply tellign the fx that we are giving SC transformed data
```

What is the Default assay now?
Check to see if the Seurat object has changed in the Assay slot at all

```{r}
DefaultAssay(pbmc.combined.sct)
pbmc.combined.sct[[]] #the orig.ident column remains in the metadata, which is how we know which data set each cell came from

```

## 3. Analysis of integrated data

Run  PCA on Integrated data

```{r}
?RunPCA
Reductions(pbmc.combined.sct) #there's no PCA on the intergrated object, these so far are only on the merged object

pbmc.combined.sct <- RunPCA(pbmc.combined.sct)

#repeating the command tells us we now have a pca reduction
Reductions(pbmc.combined.sct)
```

How many PCs should we use going forward?
```{r}
ElbowPlot(pbmc.combined.sct, ndims = 50)
#looks like 17 is a great number!
```

Visualise PCA, how does this compare to the merged data?

```{r}
?DimPlot
p1 <- DimPlot(pbmc.combined.sct, reduction = "pca") + 
  ggtitle("Integrated")
#here's the pca of intergrated data
#let's show pca of merged data next to it

p2 <- DimPlot(seurat_merged, reduction = "pca")+ 
  ggtitle("Merged")

cowplot::plot_grid(p2, p1)
```



Run UMAP 
```{r}
?RunUMAP
DefaultAssay(pbmc.combined.sct) #the default assay is still the integrated one that we created, called v2_v3_integrated

pbmc.combined.sct <- RunUMAP(pbmc.combined.sct, dims = 1:17, reduction = "pca")


```

```{r}
#just plotting again but with the umap reduction method instead of the pca reduction method

p1 <- DimPlot(pbmc.combined.sct, reduction = "umap") + 
  ggtitle("Integrated")
#here's the pca of intergrated data
#let's show pca of merged data next to it

p2 <- DimPlot(seurat_merged, reduction = "umap")+ 
  ggtitle("Merged")

cowplot::plot_grid(p2, p1)
```


```{r}
DefaultAssay(pbmc.combined.sct)
```

Find neighbors and clusters for the integrated data

Run `FindNeighbors()` and `FindClusters()` 
Try using multiple resolutions at the same time for the clustering e.g. 0.5 and 0.8
What other clustering algorithms could be used here? Which one is used by default?


```{r}
?FindNeighbors
pbmc.combined.sct <- FindNeighbors(pbmc.combined.sct, reduction = "pca", dims=1:20)

pbmc.combined.sct <- FindClusters(pbmc.combined.sct, resolution = c(0.5, 0.8, 1))

```


How many clusters have been defined at each resolution?
```{r}
pbmc.combined.sct[[]]

#isolate a column using either of these two methods: 
pbmc.combined.sct[[v2_v3_integrated_snn_res.0.5]]
pbmc.combined.sct@meta.data$v2_v3_integrated_snn_res.0.5

levels(pbmc.combined.sct$v2_v3_integrated_snn_res.0.5) #9 levels
levels(pbmc.combined.sct$v2_v3_integrated_snn_res.0.8) #11 levels
levels(pbmc.combined.sct$v2_v3_integrated_snn_res.1) #11 levels
length(levels(pbmc.combined.sct$v2_v3_integrated_snn_res.1))
# or you can do this just to get the total lenght

table(pbmc.combined.sct$v2_v3_integrated_snn_res.1) #this also gives you the number assigned to each cluster.


```

Visualse UMAP plot for batch effect and clusters

check the Default Assay to be sure

```{r}
DefaultAssay(pbmc.combined.sct)

?DimPlot()

DimPlot(pbmc.combined.sct, reduction = "umap", group.by = c("v2_v3_integrated_snn_res.0.8", "v2_v3_integrated_snn_res.0.5"))

#nb, we can't do this with the dataset that we didn't integrate, because we haven't run a find neighbours and find clusters function on it

```

Plot some canonical PBMC cluster markers e.g. S100A9, S100A8, LYZ, IGKC, GNLY, IGHM, NKG7, IGLC2

Which assay should we plot them on? (try both Integrated and SCT, is there a difference?)
First plot on the Integrated assay

```{r, fig.width=8}
?FeaturePlot
DefaultAssay(pbmc.combined.sct) <- "v2_v3_integrated"
FeaturePlot(pbmc.combined.sct, features = c("CD3D", "CD3E", "GNLY", "LYZ", "S100A8", "S100A9", "CD79A", "CD79B", "MS4A1"), reduction = "umap")

DefaultAssay(pbmc.combined.sct) <-  "SCT"
FeaturePlot(pbmc.combined.sct, features = c("CD3D", "CD3E", "GNLY", "LYZ", "S100A8", "S100A9", "CD79A", "CD79B", "MS4A1"), reduction = "umap")


#each assay has it own set of gene expression values. we will try on just the integrated assay, RNA, SCT, integrated. 

pbmc.combined.sct


```

Set default assay to "SCT" then do the same plot
```{r}
DefaultAssay()
FeaturePlot()

#done above
```

## 4. Identify markers

We want to identify markers that distinguish each cluster. 
Look up `FindMarkers` and `FindAllMarkers`, which one should we use here?

Why are we changing the DefaultAssay?
Do we need to renormalise RNA Assay?

Use the "LR" test

```{r}
DefaultAssay(pbmc.combined.sct) <- "RNA" #this is the assay that we want to use to find the markers.

#at this point, we want to specify which set of clusters we want to use (bc we generated clusters with a few different resolutions) 
#?Idents
Idents(pbmc.combined.sct) <- pbmc.combined.sct$v2_v3_integrated_snn_res.0.5

pbmc.combined.sct[[]]


?FindMarkers #Finds markers (differentially expressed genes) for identity classes
?FindAllMarkers #Finds markers (differentially expressed genes) for each of the identity classes in a dataset
#we want to do find all, bc we want to compare each cluster to all other clusters. 

#to get clusters and make umaps look good, need to remove batch effects. but now we're comparing clusters, we've already accounted for batch effects, so we now just to use raw data because it is the least manipulated. integrated assay is 2'000 genes by all cells. sct is less than the total . RNA is ALL genes

pbmc.combined.sct

#to identify: 
?FindAllMarkers 
allmarkers <- FindAllMarkers(pbmc.combined.sct, latent.vars = "orig.ident", test.use = "LR")
#Note, the default text (Wilcox) will not work because we're regressing out a variable (batch)(which is given in orig.ident)
```

Filter the top 5 markers for each cluster
use `dply::group_by` and `slice_max()`
How do you define "top"?

```{r}
allmarkers

pbmc.combined.sct.markers.top5 <- allmarkers %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n=5)


pbmc.combined.sct.markers.top5
```

plot Heatmap or Dotplot of markers

try functions `Seurat::DoHeatmap` or `Seurat::Dotplot`

```{r}
?Seurat::DoHeatmap

DefaultAssay(pbmc.combined.sct) <-  "SCT" #
Seurat::DoHeatmap(pbmc.combined.sct, features = pbmc.combined.sct.markers.top5$gene)

#need to wrap gene list in a uniq function, otherwise it will cause errors here. 
features_unique = unique(pbmc.combined.sct.markers.top5$gene)
Seurat::DotPlot(pbmc.combined.sct, features = features_unique)+ 
  theme(axis.text.x = element_text(angle=90))
```

How well do you think the clustering has worked?
Are you able to identify each cluster?
Would you re-evaluate any of the clusters?


## Optional 

Spend some time experimenting with different statistical tests in the FindAllMarkers command
Try some different parameters
Sort by log2fc vs padj
How does this affect the plots?
Use your results to label the identity of each cluster and plot on a UMAP

# Bonus Exercise: Run Harmony based integration

You can run Harmony within your Seurat workflow. You'll only need to make two changes to your code.

IMPORTANT DIFFERENCE: In the Seurat integration tutorial, you need to define a Seurat object for each dataset. 
With Harmony integration, create only one Seurat object with all cells.
Before running Harmony, make a Seurat object and following the standard pipeline through PCA.

After that Run Harmony with the `RunHarmony()` function
In downstream analyses, use the Harmony embeddings instead of PCA.
For example, run Harmony and then UMAP in two lines.

Vignette: http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/SeuratV3.html


Create a merged object, run SCTransform and PCA
```{r}
seurat_merged <- merge()

#not necessary right now as we have a seurat object with everything
seurat_merged
#printing the object at this points shows us that everything is okay
```


Run Harmony and include the optimisation graph

```{r}
seurat_merged_harmony <- RunHarmony(seurat_merged, group.by.vars = "orig.ident", reduction = "pca", dims.use = 1:17, plot_convergence = TRUE)

#the outcome here is showing us that, there is a dramatic improvement after 20 iteractions
```


Run UMAP, FindNeighbours and FindClusters, then visualise some key genes. 
```{r}
seurat_merged_harmony
Reductions(seurat_merged_harmony)


seurat_merged_harmony <- RunUMAP(seurat_merged_harmony, reduction = "harmony", dims = 1:30)
seurat_merged_harmony <- FindNeighbors(seurat_merged_harmony, reduction = "harmony", dims = 1:30)
seurat_merged_harmony <- FindClusters(seurat_merged_harmony)
```

How do the Harmony and Seurat integration methods compare?

```{r}
p1 <- DimPlot(seurat_merged_harmony, reduction = "umap", group.by = "orig.ident") + 
  ggtitle("Harmony")

p2 <- DimPlot(pbmc.combined.sct, reduction = "umap", group.by = "orig.ident") + 
  ggtitle("Integrated")

cowplot::plot_grid(p1, p2)



p1 <- DimPlot(seurat_merged_harmony, reduction = "umap") + 
  ggtitle("Harmony")

p2 <- DimPlot(pbmc.combined.sct, reduction = "umap") + 
  ggtitle("Integrated")

cowplot::plot_grid(p1, p2)
```


Compute LISI : https://github.com/immunogenomics/LISI 


