---
title: "DESeq2 Tutorial"
author: "Charlie George & Modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources

-   **DESeq2 Vignette** - <http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>
-   **Bioconductor RNAseq Workflow Book** - <http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html>

# install packages

-   If i didn't already have these installed in my environment - I would delete the \# on lines 19-23 and 26-39 and run the code

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

-   what do these packages do? Add comments to remind you what they do

```{r}
library("pheatmap")   
library("RColorBrewer") #color palette generation package, e.g. can give to GG plot 
library(tidyverse)     
library(DESeq2)      
library(ashr)         # 
library(apeglm)      #
library(org.Mm.eg.db) #
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

-   We made this (or a very similar version) using featurecounts in week1 on the cluster
-   raw data -\> fastqc -\> hisat2 -\> featurecounts

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames(var = "Geneid") %>% #bc we want the geneid values to become column titles 
                  dplyr::select(starts_with("ERR")) %>% #samples will be named differently in different datasets. you might need to use ends_with to select, or even use a line like: "dplyr::select(!c(Chr|Start|End|Strand|Length))" %>% to remove 1 by 1
                  `colnames<-`(str_remove(colnames(.), ".bam")) %>% 
                  as.matrix()



# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 
dim(counts_table)
dim(gene_info)
```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv")
sample_info

sample_info <- column_to_rownames(sample_info, "sample_id") #need to rename the column names as sample_id otherwise the check identical in the next box won't work

# What class is the cell_type currently? 
#ANS running the sample_info on its own gives a preview. the data is in a tibble. the class of cell type is currently character. 

# We want to change it to a factor with 2 levels (the hypothesis we will test)
sample_info$cell_type <- factor(sample_info$cell_type, levels=c("CD4", "CD8"))

# Which should be our reference level? 
#everything else will be a fold change relative to your base level. the base level is the first level in your factor, here CD4


# what other columns might we want to include in our design that we want to change to factors? 
#genotype and replicate we would like as a factor. also we need to convert day_collected as a factor. 
sample_info$genotype <- factor(sample_info$genotype) #if you don't specify levels = ,it does it automatically. 
sample_info$replicate <- factor(sample_info$replicate)
sample_info$day_collected <- factor(sample_info$day_collected)


```

# Check the sample names are the same and in the same order in the sample_info and counts_table

```{r}

# Column names of the count table and row names of the sample info table must be in the same order
# Check this 


colnames(counts_table) == rownames(sample_info)

# make sample_info row names from sample_id 

# remove .bam from colnames 


```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(
  countData = counts_table, 
  colData = sample_info, 
  design = ~ cell_type
)
dds

  #design = ~ genotype + replicate + cell_type

```

## Investigate the dds object

```{r}
# what does it look like? - e.g. use str(dds)
str(dds)

# how to access counts
assays(dds)
assays(dds)$counts

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds) 
#NB you can incorporate the geneinfo data here. 
#rowData(dds) = gene_info 

# access the design
design(dds)

# does it have any rowRanges()?
rowRanges(dds) 

# what are its rownames and colnames?
rownames(dds)
colnames(dds)

# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object

```{r}
# practice subsetting dds object
# select first 3 columns
dds[, 1:3]


```

# Filter out unexpressed genes

-   It's good practice to filter out genes with 0 counts across all samples at the start
-   This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis
-   Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.

```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`



dds_filtered <- dds[rowSums(counts(dds)) > 0,] #we're selecting columns based on this boolean

dds

rowSums(counts(dds)) > 0 

# check dims of dds_filtered  
  
  
```

# Perform exploratory analysis on dds_filtered object

1.  transform dds_filtered object using `rlog()` functions
    -   name your new variables `rlog_transformed_dds`
    -   we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    -   check the class of the rlog_transformed_dds objects
    -   you can look at the tranformed counts table using `assay()`

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!

```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind=TRUE)
rlog_transformed_dds


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds) #assay is showing you the data. 

```

# Optional: If you have time do the tranformation again but using the vst() function

-   As a guide
    -   rlog - more robust than VST when size factors vary widely (10x)
    -   VST - less sensitive to high count outliers than rlog, quicker - recommended most often

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- vst(dds_filtered, blind=TRUE)

```

# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()

```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 

#firt look at it again
assay(rlog_transformed_dds)

#now make correlation calculation, for the assay in here.
cor_matrix <- cor(assay(rlog_transformed_dds))
cor_matrix

# use pheatmap to plot the correlation matrix 
pheatmap(cor_matrix)

#what is this showing? it shows how gene expression values are correlated between different samples. between teh cell types that are the same, it makes sense that they're very well correlated. 


#can play with this 
pheatmap(cor_matrix, 
         color = (colorRampPalette((brewer.pal(n=5, name = "Reds"))))(150))

#NB colour = rev(colorRampPalette((brewer.pal(n=5, name = "Reds"))))(100))
#to reverse the colors

```

# try ploting a heatmap of sample distances

```{r}
# get sample distances -> Annotated what this code is doing?

#to show whats going on in steps: 
# this first line just shows what's there
(assay(rlog_transformed_dds))[1:8, 1:8]

#then we transpose it because we need it in this format for the distance fx
t(assay(rlog_transformed_dds))[1:8, 1:8]

#transformed it because we want distance between samples, not distance between genes. this is bc dist calculates the distance between samples across rows(genes), which gives you the important information you're after
sampleDists <- dist(t(assay(rlog_transformed_dds)))
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255) #the 255 gives us the number of colors that will make u pteh matrix. so if it is 3, you get only three colors on the legend. larger values will give a nice smoooth 

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data?

Answer: samples 89 is a duplicate of 83, sample 82 is a duplicate of 88

## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6] #could also use -2? or give exact column names
rlog_transformed_dds_subset <- rlog_transformed_dds[, -c(7,8)]
rlog_transformed_dds_subset <- rlog_transformed_dds[, -(7:8)]

rlog_transformed_dds_subset
assay(rlog_transformed_dds_subset) 
# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above


```

What do you notice now about the data and the clustering? They are now clustering into two groups, based on the cell types.

# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r}
# plot pca of rlog_transformed_dds using plotPCA()
colData(rlog_transformed_dds_subset) #use coldata to see what interesting groups you might want to put in there. 

DESeq2::plotPCA(rlog_transformed_dds_subset, intgroup = 'cell_type') +
  coord_cartesian() #this line is to make sure that the graph fills the space
  

# what does the intgroup parameter do? have a go at changing it...
DESeq2::plotPCA(rlog_transformed_dds_subset, intgroup = 'replicate') +
  coord_cartesian()

# how many variable genes is it using for pca? 
# ntop is 500 as the default. 

# plot the pca for all genes 
DESeq2::plotPCA(rlog_transformed_dds_subset, intgroup = 'cell_type', ntop = nrow(rlog_transformed_dds_subset)) +
  coord_cartesian() #this line is to make sure that the graph fills the space
```

What does it tell us about our sample?

Answer:there is more variation between cells than between replicates

------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis e.g. - update design with variables in sample info table we want to regress out - remove samples (poor quaility or those with name switches etc)

# Revisiting our inital data

What do we need to do to our data???

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}
dds_filtered
# do we need to remove/edit samples? 
dds_filtered_updated <- dds_filtered[,1:6]

# do we want to update the design? Is this nessecary in this case?
design(dds_filtered_updated)
# e.g. design(dds) <- ~replicate + condition


#check up on it
dds_filtered_updated
assays(dds_filtered_updated) #
```

# Now we can start our actual DE!!!! yay

Run deseq2 modelling using one function (DESeq())

```{r}

# look at dds slots then run deseq2 function 
# what has happened to your dds slots?
slotNames(dds_filtered_updated)
slotNames(dds_filtered)

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

# sumarise results 
summary(results_obj)

```

Have a look at the outputs

# Now we will run the analysis step by step to see whats going on

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)
# Access the size factors from the dds_filtered_updated object
dds_filtered_updated$sizeFactor

colData(dds_filtered_updated)
# Do you see any systematic differences in size factors between groups?
#size factors give you a single factor value that tells you how much of an outlier each sample is

```

# Optional: Plot size factors as barchart using ggplot

```{r}

sample_size_factors <- as.data.frame(colData(dds_filtered_updated)) %>% 
  rownames_to_column(var = "SampleID")
sample_size_factors

ggplot(sample_size_factors, aes(x = SampleID, y=sizeFactor)) +
  geom_col(aes(fill = cell_type))

```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts
normalised_counts <- counts(dds_filtered_updated, normalized = TRUE) #don't reassign this to dds_filtered_updated, bc it returns a matrix!! 
# check the class of the output
class(normalised_counts)

```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 
dds_filtered_updated <- estimateDispersions(dds_filtered_updated)

# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts((dds_filtered_updated))
```

## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest


wald_test_output <- nbinomWaldTest(dds_filtered_updated)

wald_result <- results(wald_test_output)
summary(wald_result)
```

# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# generate dds object 
dds <- DESeqDataSetFromMatrix(
  countData = counts_table, 
  colData = sample_info, 
  design = ~ cell_type
)
dds

# filter rows with 0 counts 
dds_filtered <- dds[rowSums(counts(dds)) > 0,] #we're selecting columns based on this boolean

# remove unwanted samples 
dds_filtered_updated <- dds_filtered[,1:6]

# make DESeqDataSet (dds) object 
#can reassign the design here with design(dds_filtered_updated) <- ~replicate + cell type

# run DESeq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 
results_obj <- results(dds_filtered_updated) # extracts the results table from a dseq analysis 
results_obj
class(results_obj)

summary(results_obj) #produces a rsummary of the results object
```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM -- see vingette("DESeq2")

```{r}

```

# Plot normalised counts for ENSMUSG00000041779

checkout the 'plotCounts()\` function

```{r}
# Plot normalised counts for gene ENSMUSG00000041779
plotCounts(dds_filtered_updated, gene = 'ENSMUSG00000041779', intgroup = 'cell_type')


#proove its there?
dim(dds_filtered_updated)
dim(counts(dds_filtered_updated, normalised = TRUE))
dim(normalised_counts)
```

# Access the results table for the comparison between CD8 and CD4 T cells . You can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control

checkout `resultsNames()` & `results()`

```{r}

results(dds_filtered_updated)
resultsNames(dds_filtered_updated)


results(dds_filtered_updated, contrast = c("cell_type", "CD4", "CD8"))
resultsNames(dds_filtered_updated)
```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot -- do they look as expected? for CD8 vs CD4

```{r}

dds_results_dataframe = as.data.frame(results(dds_filtered_updated)) %>% 
  rownames_to_column(var = "SampleID")
    
dds_results_dataframe

ggplot(dds_results_dataframe) + 
  geom_histogram(aes(x = pvalue), colour = 1, fill = 3, alpha = 0.5) + 
  geom_histogram(aes(x = padj), colour = 2, fill = 4, alpha = 0.5)



```

# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(dds_filtered_updated, alpha = 0.05) #alpha is the p-value for colouring. we've changed it to below 0.5
?DESeq2::plotMA

```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -\> see the vignette -- which shall we choose?

```{r}

# call your results table  called shrunkenLFC_dds_results_obj
resultsNames(dds_filtered_updated)

#normal method
dds_filtered_updated_shrunk <-lfcShrink(dds_filtered_updated, type = "normal", coef = "cell_type_CD8_vs_CD4")
plotMA(dds_filtered_updated_shrunk, alpha = 0.05)

#ape glm method 
dds_filtered_updated_shrunk <-lfcShrink(dds_filtered_updated, type = "apeglm", coef = "cell_type_CD8_vs_CD4")
plotMA(dds_filtered_updated_shrunk, alpha = 0.05)


#ashr method 
dds_filtered_updated_shrunk <-lfcShrink(dds_filtered_updated, type = "ashr", coef = "cell_type_CD8_vs_CD4")
plotMA(dds_filtered_updated_shrunk, alpha = 0.05)



dds_filtered_updated_shrunk
```

# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 

# How many Ensembl IDs are not assigned an gene symbol and how many gene symbols are duplicated?



keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI

# add mapId vectors to  shrunkenLFC_dds_results_df
```

# Export all the results to CSV

```{r}
# dds_filtered_updated_shrunk <-lfcShrink(dds_filtered_updated, type = "normal", coef = "cell_type_CD8_vs_CD4")

shrunkenLFC_dds_results_df <-  as.data.frame(dds_filtered_updated_shrunk) %>% 
  rownames_to_column(var = 'gene')

shrunkenLFC_dds_results_df


write_csv(shrunkenLFC_dds_results_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```

# Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```

Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account.

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....
